syntax = "proto2";

message Coordinates {
  message RightAscention {
    required int32 hours = 1;
    required int32 minutes = 2;
    required float seconds = 3;
  }

  message Declination {
    required int32 degrees = 1;
    required int32 minutes = 2;
    required float seconds = 3;
  }

  required RightAscention right_ascention = 1;
  required Declination declination = 2;
}

message Officer {
  required string first_name = 1;
  required string last_name = 2;
  required string rank = 3;
}

message Spaceship {
  enum Alignment {
    Ally = 0;
    Enemy = 1;
  }

  required Alignment alignment = 1;
  required string name = 2;
  enum Class {
    Corvette = 0;
    Frigate = 1;
    Cruiser = 2;
    Destroyer = 3;
    Carrier = 4;
    Dreadnought = 5;
  }
  required Class ship_class = 3;
  required float length = 4;
  required int32 crew_size = 5;
  required bool armed = 6;
  repeated Officer officers = 7;
}

service SpaceshipService {
  rpc GetSpaceshipInfo (Coordinates) returns (stream Spaceship);
}
